cardano-crypto-class/src/Cardano/Crypto/DSIGN/BLS12381.hs

```diff
@@ -0,0 +1,358 @@
+import GHC.TypeNats (KnownNat)
+import NoThunks.Class (NoThunks)
+import System.IO.Unsafe (unsafeDupablePerformIO)
+
+data BLS12381DSIGN curve
+
+-- Making sure different 'Signature schemes are not 'Coercible', which would ruin the
+-- intended type safety:
+type role BLS12381DSIGN nominal
+
+-- | Two versions of BLS12-381 DSIGN: one optimized for minimal verification key size,
+-- the other for minimal signature size.
+type BLS12381MinVerDSIGN = BLS12381DSIGN Curve1
+
+type BLS12381MinSigDSIGN = BLS12381DSIGN Curve2
+
+type family CurveVariant (c :: Type) :: Symbol where
+  CurveVariant Curve1 = "min-Public-Key"
+  CurveVariant Curve2 = "min-Signature"
+
+-- VerKeyDSIGN uses PointSize curve
+deriving newtype instance
+  BLS curve =>
+  Eq (VerKeyDSIGN (BLS12381DSIGN curve))
+
+-- SigDSIGN uses PointSize (Dual curve)
+deriving newtype instance
+  BLS (Dual curve) =>
+  Eq (SigDSIGN (BLS12381DSIGN curve))
+
+type BLS12381Ctx curve =
+  ( BLS curve
+  , BLS (Dual curve)
+  , FinalVerifyOrder curve
+  , KnownSymbol (CurveVariant curve)
+  , Typeable curve
+  )
+
+instance
+  ( BLS12381Ctx curve
+  , KnownNat (PointSize curve)
+  , KnownNat (PointSize (Dual curve))
+  ) =>
+  DSIGNAlgorithm (BLS12381DSIGN curve)
+  where
+  type SeedSizeDSIGN (BLS12381DSIGN curve) = BLST_SCALAR_SIZE
+  type SizeSignKeyDSIGN (BLS12381DSIGN curve) = BLST_SCALAR_SIZE
+  type SizeVerKeyDSIGN (BLS12381DSIGN curve) = PointSize curve
+  type SizeSigDSIGN (BLS12381DSIGN curve) = PointSize (Dual curve)
+  type Signable (BLS12381DSIGN curve) = SignableRepresentation
+
+  -- \| Context can hold domain seperation tag and/or augmentation data for signatures
+  type ContextDSIGN (BLS12381DSIGN curve) = (Maybe ByteString, Maybe ByteString)
+  type KeyGenContextDSIGN (BLS12381DSIGN curve) = Maybe ByteString
+  newtype VerKeyDSIGN (BLS12381DSIGN curve)
+    = VerKeyBLS12381 (PinnedSizedBytes (SizeVerKeyDSIGN (BLS12381DSIGN curve)))
+    deriving newtype (NFData)
+    deriving stock (Show, Generic)
+    deriving anyclass (NoThunks)
+  newtype SignKeyDSIGN (BLS12381DSIGN curve)
+    = SignKeyBLS12381 (PinnedSizedBytes (SizeSignKeyDSIGN (BLS12381DSIGN curve)))
+    deriving newtype (Eq, NFData)
+    deriving stock (Show, Generic)
+    deriving anyclass (NoThunks)
+  newtype SigDSIGN (BLS12381DSIGN curve)
+    = SigBLS12381 (PinnedSizedBytes (SizeSigDSIGN (BLS12381DSIGN curve)))
+    deriving newtype (NFData)
+    deriving stock (Show, Generic)
+    deriving anyclass (NoThunks)
+  algorithmNameDSIGN _ = "bls12-381-" ++ symbolVal (Proxy @(CurveVariant curve))
+  {-# NOINLINE deriveVerKeyDSIGN #-}
+  deriveVerKeyDSIGN (SignKeyBLS12381 skPsb) = do
+    VerKeyBLS12381 $ unsafeDupablePerformIO . psbUseAsCPtr skPsb $ \skp ->
+      psbCreate $ \vkPtp ->
+        c_blst_sk_to_pk @curve (PointPtr vkPtp) (ScalarPtr skp)
+  {-# NOINLINE signDSIGN #-}
+  signDSIGN ctx@(dst, aug) msg (SignKeyBLS12381 skPsb) =
+    warnIfNoDst ctx $ SigBLS12381 $ unsafeDupablePerformIO . psbCreate $ \sigPts -> do
```

Comment:
@lehins, wdyt of doing something like this `warnIfNoDst`? Is there a nice way of doing this? Similarly for `proveProofOfPossesion` down below.

As the warning mentions, it is really unsafe to not use a domain separation tag. That said, the IETF draft assumes no, so we do need to support it to be compatible. Of course we can also just add a comment about it, but I wonder if there is more to push this in the face of the developer.

Thanks!